// @better-fetch/fetch/dist/index.cjs
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/index.ts
var src_exports = {};
__export(src_exports, {
  betterFetch: () => betterFetch,
  createFetch: () => createFetch,
  default: () => src_default
});
module.exports = __toCommonJS(src_exports);

// src/error.ts
var BetterFetchError = class extends Error {
  constructor(message, error) {
    super(message);
    this.message = message;
    this.error = error;
  }
};
var FetchError = class extends BetterFetchError {
  constructor(status, statusText, error) {
    super(statusText);
    this.status = status;
    this.statusText = statusText;
    this.error = error;
  }
};

// src/utils.ts
var JSON_RE = /^application\/(?:[\w!#$%&*.^`~-]*\+)?json(;.+)?$/i;
function detectResponseType(request) {
  const _contentType = request.headers.get("content-type");
  const textTypes = /* @__PURE__ */ new Set([
    "image/svg",
    "application/xml",
    "application/xhtml",
    "application/html"
  ]);
  if (!_contentType) {
    return "json";
  }
  const contentType = _contentType.split(";").shift() || "";
  if (JSON_RE.test(contentType)) {
    return "json";
  }
  if (textTypes.has(contentType) || contentType.startsWith("text/")) {
    return "text";
  }
  return "blob";
}
function isJSONParsable(value) {
  try {
    JSON.parse(value);
    return true;
  } catch (error) {
    return false;
  }
}
function isJSONSerializable(value) {
  if (value === void 0) {
    return false;
  }
  const t = typeof value;
  if (t === "string" || t === "number" || t === "boolean" || t === null) {
    return true;
  }
  if (t !== "object") {
    return false;
  }
  if (Array.isArray(value)) {
    return true;
  }
  if (value.buffer) {
    return false;
  }
  return value.constructor && value.constructor.name === "Object" || typeof value.toJSON === "function";
}
function jsonParse(text) {
  if (!isJSONSerializable(text)) {
    return text;
  }
  try {
    return JSON.parse(text);
  } catch (error) {
    return text;
  }
}
function isFunction(value) {
  return typeof value === "function";
}
function getFetch(customFetchImpl) {
  if (customFetchImpl) {
    return customFetchImpl;
  }
  if (typeof globalThis !== "undefined" && isFunction(globalThis.fetch)) {
    return globalThis.fetch;
  }
  if (typeof window !== "undefined" && isFunction(window.fetch)) {
    return window.fetch;
  }
  throw new Error("No fetch implementation found");
}

// src/index.ts
var import_zod = require("zod");
var betterFetch = async (url, options) => {
  const fetch2 = getFetch(options?.customFetchImpl);
  const controller = new AbortController();
  const signal = controller.signal;
  for (const plugin of options?.plugins || []) {
    const pluginRes = await plugin(url.toString(), options);
    url = pluginRes.url;
    options = pluginRes.options;
  }
  const _url = new URL(`${options?.baseURL ?? ""}${url.toString()}`);
  const headers = new Headers(options?.headers);
  const shouldStringifyBody = options?.body && isJSONSerializable(options.body) && (!headers.has("content-type") || headers.get("content-type") === "application/json") && typeof options?.body !== "string";
  if (shouldStringifyBody) {
    !headers.has("content-type") && headers.set("content-type", "application/json");
    !headers.has("accept") && headers.set("accept", "application/json");
  }
  const query = options?.query;
  if (query) {
    for (const [key, value] of Object.entries(query)) {
      _url.searchParams.append(key, String(value));
    }
  }
  const _options = {
    signal,
    ...options,
    body: shouldStringifyBody ? JSON.stringify(options?.body) : options?.body ? options.body : void 0,
    headers,
    method: options?.method?.length ? options.method : options?.body ? "POST" : "GET"
  };
  if ("pipeTo" in _options && typeof _options.pipeTo === "function" || typeof options?.body?.pipe === "function") {
    if (!("duplex" in _options)) {
      _options.duplex = "half";
    }
  }
  const context = {
    request: new Request(_url.toString(), _options),
    options: _options,
    controller
  };
  let abortTimeout;
  if (!context.request.signal && context.options.timeout) {
    abortTimeout = setTimeout(
      () => controller.abort(),
      context.options.timeout
    );
  }
  await options?.onRequest?.(context);
  const response = await fetch2(_url.toString(), _options);
  const responseContext = {
    response
  };
  if (abortTimeout) {
    clearTimeout(abortTimeout);
  }
  await options?.onResponse?.(responseContext);
  const hasBody = response.body && context.options.method !== "HEAD";
  if (!hasBody) {
    await options?.onSuccess?.(responseContext);
    return {
      data: {},
      error: null
    };
  }
  const responseType = detectResponseType(response);
  if (response.ok) {
    if (responseType === "json" || responseType === "text") {
      const parser2 = options?.jsonParser ?? jsonParse;
      const text2 = await response.text();
      const json = await parser2(text2);
      await options?.onSuccess?.(responseContext);
      return {
        data: json,
        error: null
      };
    } else {
      return {
        data: await response[responseType](),
        error: null
      };
    }
  }
  await options?.onError?.(responseContext);
  if (options?.retry) {
    await options?.onRetry?.(responseContext);
    return await betterFetch(url, {
      ...options,
      retry: options.retry - 1
    });
  }
  const parser = options?.jsonParser ?? jsonParse;
  const text = await response.text();
  const errorObject = isJSONParsable(text) ? await parser(text) : text ? {
    message: text
  } : void 0;
  if (options?.throw) {
    throw new FetchError(response.status, response.statusText, errorObject);
  }
  if (errorObject) {
    return {
      data: null,
      error: {
        ...errorObject,
        status: response.status,
        statusText: response.statusText
      }
    };
  }
  return {
    data: null,
    error: {
      ...{},
      status: response.status,
      statusText: response.statusText
    }
  };
};
var createFetch = (config) => {
  const $fetch = async (url, options) => {
    return await betterFetch(url, {
      ...config,
      ...options
    });
  };
  $fetch.native = fetch;
  return $fetch;
};
betterFetch.native = fetch;
var src_default = betterFetch;
var routes = {
  "/": {
    output: import_zod.z.object({
      message: import_zod.z.string()
    })
  },
  "/signin": {
    input: import_zod.z.object({
      username: import_zod.z.string(),
      password: import_zod.z.string()
    }),
    output: import_zod.z.object({
      token: import_zod.z.string()
    })
  },
  "/signup": {
    input: import_zod.z.object({
      username: import_zod.z.string(),
      password: import_zod.z.string(),
      optional: import_zod.z.optional(import_zod.z.string())
    }),
    output: import_zod.z.object({
      message: import_zod.z.string()
    })
  },
  "/query": {
    query: import_zod.z.object({
      term: import_zod.z.string()
    })
  }
};
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  betterFetch,
  createFetch
});
